sklearn/_fast_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,757 @@
+        self.min_samples_leaf = min_samples_leaf
+        self.l2_regularization = l2_regularization
+        self.max_bins = max_bins
+        self.n_iter_no_change = n_iter_no_change
+        self.validation_fraction = validation_fraction
+        self.scoring = scoring
+        self.tol = tol
+        self.verbose = verbose
+        self.random_state = random_state
+
+    def _validate_parameters(self):
+        """Validate parameters passed to __init__.
+
+        The parameters that are directly passed to the grower are checked in
+        TreeGrower."""
+
+        if self.loss not in self._VALID_LOSSES:
+            raise ValueError(
+                "Loss {} is not supported for {}. Accepted losses: "
+                "{}.".format(self.loss, self.__class__.__name__,
+                             ', '.join(self._VALID_LOSSES)))
+
+        if self.learning_rate <= 0:
+            raise ValueError('learning_rate={} must '
+                             'be strictly positive'.format(self.learning_rate))
+        if self.n_estimators < 1:
+            raise ValueError('n_estimators={} must not be smaller '
+                             'than 1.'.format(self.n_estimators))
+        if self.n_iter_no_change is not None and self.n_iter_no_change < 0:
+            raise ValueError('n_iter_no_change={} must be '
+                             'positive.'.format(self.n_iter_no_change))
+        if (self.validation_fraction is not None and
+                self.validation_fraction <= 0):
+            raise ValueError(
+                'validation_fraction={} must be strictly '
+                'positive, or None.'.format(self.validation_fraction))
+        if self.tol is not None and self.tol < 0:
+            raise ValueError('tol={} '
+                             'must not be smaller than 0.'.format(self.tol))
+
+    def fit(self, X, y):
+        """Fit the gradient boosting model.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+            The input samples.
+
+        y : array-like, shape=(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+
+        fit_start_time = time()
+        acc_find_split_time = 0.  # time spent finding the best splits
+        acc_apply_split_time = 0.  # time spent splitting nodes
+        # time spent predicting X for gradient and hessians update
+        acc_prediction_time = 0.
+        X, y = check_X_y(X, y, dtype=[X_DTYPE])
+        y = self._encode_y(y)
+        rng = check_random_state(self.random_state)
+
+        self._validate_parameters()
+        self.n_features_ = X.shape[1]  # used for validation in predict()
+
+        # we need this stateful variable to tell raw_predict() that it was
+        # called from fit(), which only passes pre-binned data to
+        # raw_predict() via the scorer_ attribute. predicting is faster on
+        # pre-binned data.
+        self._in_fit = True
```

Comment:
@ogrisel WDYT of this?

In pygbm we get around this by checking the dtype of `X` in `raw_predict` but we can't do that in sklearn right?